ddoc_preprocessor.d

```diff
@@ -194,3 +195,61 @@ auto genHeader(string fileText)
     return updateDdocTag(fileText, ddocKey, newContent);
 }
 
+// parse the menu from the Ddoc file
+auto specTocEntries()
+{
+    alias Entry = Tuple!(string, "name", string, "title", string, "fileName");
+    Entry[] entries;
+
+    static immutable specDir = __FILE_FULL_PATH__.dirName.buildNormalizedPath("spec");
+    static immutable mainFile = specDir.buildPath("./spec.ddoc");
+
+    auto specText = mainFile.readText;
+    if (!specText.findSkip("SUBMENU2"))
+        writeln("Menu file has an invalid format.");
+    foreach (line; specText.splitter("\n"))
+    {
+        enum ddocEntryStart = "$(ROOT_DIR)spec/";
+        if (line.find!(not!isWhite).startsWith(ddocEntryStart))
+        {
+            auto ps = line.splitter(ddocEntryStart).dropOne.front.splitter(",");
+            auto name = ps.front.stripExtension.withExtension(".dd").to!string;
+            auto fileName = specDir.buildPath(name);
+            auto title = ps.dropOne.front.idup.strip;
+            entries ~= Entry(name, title, fileName);
+        }
+    }
+    return entries;
+}
+
+// Automatically generate spec/grammar
+auto genGrammar(string fileText)
+{
+    import std.uni : toLower;
+
+    enum ddocKey = "$(GRAMMAR_SUMMARY";
+    auto newContent = ddocKey ~ "\n";
+
+    if (fileText.canFind(ddocKey))
+    {
+        foreach (i, entry; specTocEntries)
+        {
+            if (entry.fileName.endsWith("grammar.dd", "lex.dd", "simd.dd"))
```

Comment:
- `sim.dd` defines this:

```
$(GRAMMAR
$(I typeNN)
)
```

- `lex.dd` is excluded in the current spec

![image](https://user-images.githubusercontent.com/4370550/35479040-040a111a-03ed-11e8-8a05-72e1a3c602d0.png)